Configuration and event updates

Dando-Real-ITA opened the pull request:
- Started to migrate some config from MACD bot
- Simplified required config in VENUES
- Executing fetch only once and start, and using events to update risk and orders
- Added orders tracking support
- Reduced candles in memory to 10 times LOOKBACK to avoid out of memory on long runnin bots
- Added support for API subaccounts
- Implemented order management
- Not creating new orders that would be cancelled with ReduceOnly
- Created a generic place_orders for bids/asks
- Removed VENUE_IDS
- Added Richard code for google sheet
- Added Top of Book mode
- Fixed INIT state
- Added retry for all call_endpoint

Dando-Real-ITA committed:
Inverted cancel_order and orders_intent

Dando-Real-ITA committed:
Made multiplier optional per-contract

Dando-Real-ITA committed:
Merge remote-tracking branch 'origin/main' into main

Dando-Real-ITA committed:
Configuration and event updates

- Started to migrate some config from MACD bot
- Simplified required config in VENUES
- Executing fetch only once and start, and using events to update risk and orders
- Added orders tracking support
- Reduced candles in memory to 10 times LOOKBACK to avoid out of memory on long runnin bots

Dando-Real-ITA committed:
Merge remote-tracking branch 'origin/main' into main

Dando-Real-ITA committed:
More updates for grid bot

- Added support for API subaccounts
- Implemented order management
- Not creating new orders that would be cancelled with ReduceOnly

Dando-Real-ITA commented:
This PR now requires to set correctly variables MAIN_VENUE and VENUE_IDS

Thomas1Bunting commented on src/grid-bots/multi-vendor-multi-contract.py:
```diff
@@ -7,146 +7,106 @@
+TIME_LOOKUP = {
+  '1m':  60_000,
+  '5m' : 60_000 * 5,
+  '15m': 60_000 * 15,
```
15m won't be supported soonish so could remove.

Dando-Real-ITA replied:
Removed in https://github.com/profitviews/bitmex-algo-academy/pull/4/commits/ced14572d447ca115d20d87648e169e9876aa4d5

Thomas1Bunting commented on src/grid-bots/multi-vendor-multi-contract.py:
```diff
@@ -7,146 +7,106 @@
+  LEVEL = '1m'
   SRC = 'bitmex'                         # exchange name as in the Glossary
+  MAIN_VENUE='account1'
+  VENUE_IDS = {
```
This isn't required

Dando-Real-ITA replied:
MAIN_VENUE is for the candles ( bitmex crashed on fetch_candles if you use a subaccount )
New code does not need VENUE_IDS since https://github.com/profitviews/bitmex-algo-academy/pull/4/commits/d4a9353d14b2aed36ec623dd64caec24ea218cf8

Thomas1Bunting commented on src/grid-bots/multi-vendor-multi-contract.py:
```diff
@@ -7,146 +7,106 @@
-          'tob': (np.nan, np.nan),
-          'max_risk': 400000*10,                  # Means max 10 times the base order size
-          'current_risk': 0,
+          'order_size': 1_000 * 400,
```
What does the 1_000 represent? I'm not too familiar with python.

Dando-Real-ITA replied:
It is same a 1000 with a "pretty" separator for thousands. It is optional.

Thomas1Bunting commented on src/grid-bots/multi-vendor-multi-contract.py:
```diff
@@ -256,47 +255,128 @@ async def update_limit_orders(self):
+                  # Amend the order
+                  # self.amend_order(venue, order_id=order_id, price=bid)
+                  try:
+                    self.call_endpoint(
```
imo, we won't HAVE to manage amends, we can still cancel and recreate the orders which will remove a lot of state management. you use only one more API request to cancel all so its not the end of the world to cancel / new

Dando-Real-ITA replied:
I personally prefer the state management, it reduces calls by a lot, also amends are faster from an API standpoint because bitmex is column based so changing price happens faster then deleting an order ( with all its data ) and recreating it

If you prefer, I can add an option and have a path that uses the cancel all/recreate path and another that uses the state management

Dando-Real-ITA committed:
Unified place_orders

- No more need for VENUE_IDS

Dando-Real-ITA committed:
Added Richard google sheet code

Dando-Real-ITA committed:
Removed 15m

Dando-Real-ITA committed:
Multiple updates

- Readded debounce
- Added TOB mode with related variables: price_spread, mode and tob_is_bid
- Added INIT var to avoid errors in callback init
- Default mode if not selected is GRID
- Added multi retry up to 10 times for each call_endpoint, and more log output
- Using debounce on update_tob_order to avoid api overload
- Using price_spread to set order not exactly at tob ( useful for reverses )
- On reverse tob, use a bigger spread ( 10 times price precision ) to facilitate setting up the limit order
- Only updating tob orders if there is a tob change on quote_update ( because can be size only )
- Reversing tob order on order completed

Dando-Real-ITA committed:
MACD merge

- Fixed amend retry, only marking order as used on success
- Fast fail on amend for Canceled and Filled orders
- Created taker_order that creates a market order and a limit order
- Created trade that select which orders to execute
- Increased tob order debounce to 5
- Updated reverse_tob_order to verify first if the order was filled and not canceled
- Added all MACD methods

Dando-Real-ITA committed:
Updated MACD to latest code

Dando-Real-ITA committed:
More updates

- Added REDUCE_TAKER option
- Added order and position resync every 30 min ( for bad states that can happen )
- Using nanmean as a mean to avoid nan results for macd

Dando-Real-ITA committed:
Removed a call

Dando-Real-ITA committed:
Merge remote-tracking branch 'origin/main' into main

Dando-Real-ITA committed:
Change defaults and stability update

- Updated some default values
- Added option to always refetch order and status at every cicle, in case there are update callback problems
